### ISSUE TITLE
关于对比学习中代码实现问题

### ISSUE THREAD

**Issue by alala521** [Glacier0012/CREC#3]
对比学习是拉进视觉特征和正样本间的距离，而在代码中
# calculate the distance between fusion feature and textual features (y: f_T+, n_y: f_T-)
f_v = self.visual_proj(pooled).view(bs, 1, -1)
f_t = torch.hstack((y.view(bs, 1, -1), n_y.view(bs, 1, -1)))
logits = torch.bmm(f_v, f_t)
logits = F.softmax(logits.view(bs, -1), dim=1)
cl_label = 1 - det_label[:, :, 4].view(-1)                  # cf_id=0 -> postive -> cl_label=1
cl_label = cl_label.to(torch.int64)
loss_cl = self.cl_loss(logits/temp, cl_label)  ##修改

我对这段代码的理解是 既然f_t=[正样本，负样本]，在做最后的交叉熵时，logits应该和0去做交叉熵，因为f_t
的0位置为正样本的位置。可是在代码中确是和cl_lable去做交叉熵。

**Comment by Glacier0012** [comment]
非常感谢指出问题。此处为代码整理时产生的bug，原始写法为：

```python
cf_id = det_label[:, :, 4].view(-1)                             # [bs]
cf_id = cf_id.to(torch.int64)
cl_label = torch.zeros(logits.shape).to(logits.device)          # [bs, 2]
for i in range(bs):
    cl_label[i][cf_id[i]] = 1.
loss_cl = self.cl_loss(logits/temp, cl_label)
```

符合对比学习损失定义。代码简化后将cl_label定义错误，造成误解十分抱歉。

另外，关于为什么用det_label计算cl_label而不是直接将cl_label写为0去求交叉熵，这是因为我们的数据集混合了**反事实**和**非反事实**的样本。对反事实样本而言，其neg是非反事实样本，cl_label对应为1，所以需要用det_label而不是0来计算contrastive loss.

总而言之非常感谢！代码已更正。
